### PR TITLE
Bug 1397913 - Allow categorical histograms in Main Summary

### PR DESCRIPTION
**Pull request by fbertsch**
Categorical histograms will show a Map[String, Integer] rather
than the Map[Integer, Integer] we've shown for all other
histograms.

### CODE REVIEW

File: src/main/scala/com/mozilla/telemetry/metrics/Histogram.scala [hunk 1]
```diff
@@ -15,6 +15,8 @@ case class CountHistogram(keyed: Boolean, originalName: String, process: Option[
 case class EnumeratedHistogram(keyed: Boolean, originalName: String, nValues: Int, process: Option[String] = None) extends HistogramDefinition
 case class LinearHistogram(keyed: Boolean, originalName: String, low: Int, high: Int, nBuckets: Int, process: Option[String] = None) extends HistogramDefinition
 case class ExponentialHistogram(keyed: Boolean, originalName: String, low: Int, high: Int, nBuckets: Int, process: Option[String] = None) extends HistogramDefinition
+case class CategoricalHistogram(keyed: Boolean, originalName: String, labels: Seq[String], process: Option[String] = None) extends HistogramDefinition
+object CategoricalHistogram{ val SpillBucketName = "spill" }
```
Comment (mreid-moz) [inline src/main/scala/com/mozilla/telemetry/metrics/Histogram.scala]: Please add a comment about this Spill Bucket

**Comment by fbertsch** [comment]
@mreid-moz I'm nervous about adding it to Longitudinal, because it's not really been asked for and may break the job (Longitudinal doesn't test that it can be created within Spark, unlike Main Summary [0]). Also, the bug comment says it should have NO effect on longitudinal :)

[0] https://github.com/mozilla/telemetry-batch-view/blob/master/src/test/scala/com/mozilla/telemetry/views/MainSummaryViewTest.scala#L1002

**Comment by mreid-moz** [comment]
Gotcha, clearly my look was too quick :)

**Comment by fbertsch** [comment]
@mreid-moz RFAL

File: src/main/scala/com/mozilla/telemetry/metrics/Histogram.scala [hunk 2]
```diff
@@ -16,6 +16,15 @@ case class EnumeratedHistogram(keyed: Boolean, originalName: String, nValues: In
 case class LinearHistogram(keyed: Boolean, originalName: String, low: Int, high: Int, nBuckets: Int, process: Option[String] = None) extends HistogramDefinition
 case class ExponentialHistogram(keyed: Boolean, originalName: String, low: Int, high: Int, nBuckets: Int, process: Option[String] = None) extends HistogramDefinition
 
+/**
+ * All non-labeled values will be aggreagted into the spill bucket,
```
Comment (mreid-moz) [inline src/main/scala/com/mozilla/telemetry/metrics/Histogram.scala]: nit: aggregated

File: src/test/scala/com/mozilla/telemetry/views/MainSummaryViewTest.scala [hunk 3]
```diff
@@ -1623,4 +1605,151 @@ class MainSummaryViewTest extends FlatSpec with Matchers{
...
+      .getValuesMap(expected.keys.toList)
+
+    for ((f, v) <- expected) {
+      withClue(s"$f:") { actual.get(f) should be (Some(v)) }
+    }
+
+    actual should be (expected)
+    spark.stop()
+  }
+
```
Comment (mreid-moz) [inline src/test/scala/com/mozilla/telemetry/views/MainSummaryViewTest.scala]: nit: pls remove these blank lines.

File: src/test/scala/com/mozilla/telemetry/views/MainSummaryViewTest.scala [hunk 4]
```diff
@@ -1623,4 +1605,151 @@ class MainSummaryViewTest extends FlatSpec with Matchers{
...
+    val sc = spark.sparkContext
+    import spark.implicits._
+
+    val message = RichMessage(
+      "1234",
+      Map(
+        "documentId" -> "foo",
+        "submissionDate" -> "1234",
+        "payload.histograms" -> """{
+    "MOCK_CATEGORICAL": {
```
Comment (mreid-moz) [inline src/test/scala/com/mozilla/telemetry/views/MainSummaryViewTest.scala]: nit: you could improve the indenting here using the `stripMargin` approach if you like.

**Pull request merged**